dvc/stage/imports.py

```diff
@@ -17,20 +17,37 @@ def _update_import_on_remote(stage, remote, jobs):
     stage.outs[0].transfer(url, odb=odb, jobs=jobs, update=True)
 
 
-def update_import(stage, rev=None, to_remote=False, remote=None, jobs=None):
+def update_import(
+    stage, rev=None, to_remote=False, remote=None, no_download=None, jobs=None
+):
     stage.deps[0].update(rev=rev)
+    outs = stage.outs
+    deps = stage.deps
+
     frozen = stage.frozen
     stage.frozen = False
+
+    if stage.outs:
+        stage.outs[0].clear()
     try:
         if to_remote:
             _update_import_on_remote(stage, remote, jobs)
         else:
-            stage.reproduce(jobs=jobs)
+            stage.reproduce(no_download=no_download, jobs=jobs)
     finally:
+        if deps == stage.deps:
+            stage.outs = outs
```

Comment:
I am not sure what this is doing here.

Reply:
A few lines above we `clear` the stage's outs (delete hash_info, meta,  and obj from the out).
This is important because when using `no_download == True`, old outputs could be retained (even if deps are changed).
If the deps did not change, we can restore the previous outs since they did not change either.

Reply:
But the `.clear()` mutates the output itself right? Isn't `stage.outs[0]` and `outs[0]` the same instance here?